### PR TITLE
Port over users API.

### PR DESCRIPTION
**Pull request by ketan**
Merge this after #1260.

### CODE REVIEW

**Comment by ketan** [comment]
PR for api docs - gocd/api.go.cd#6

**Comment by zabil** [comment]
I noticed that a bad request  (400) , if you send malformed json returns as an html and not json.

File: server/test/integration/com/thoughtworks/go/server/service/UserServiceIntegrationTest.java [hunk 1]
```diff
@@ -70,6 +71,7 @@
 import static org.junit.Assert.assertThat;
 import static org.junit.Assert.fail;
 import static org.junit.matchers.JUnitMatchers.containsString;
+import static org.mockito.Mockito.verify;
```
Comment (zabil) [inline server/test/integration/com/thoughtworks/go/server/service/UserServiceIntegrationTest.java]: The verify shows up as unused

File: server/test/integration/com/thoughtworks/go/server/service/UserServiceIntegrationTest.java [hunk 2]
```diff
@@ -609,6 +611,73 @@ public void shouldDisableAdminSelectionWhenUserIsMemberOfAdminRole() throws Exce
         assertThat(userService.getAdminAndRoleSelections(Arrays.asList("foo", "quux")).getAdminSelection(), is(new TriStateSelection(Admin.GO_SYSTEM_ADMIN, TriStateSelection.Action.add, false)));
     }
 
+    @Test
+    public void updateShouldUpdateEnabledState() throws Exception {
```
Comment (zabil) [inline server/test/integration/com/thoughtworks/go/server/service/UserServiceIntegrationTest.java]: Can we stick to the convention of "shouldUpdateUser..."

File: server/test/integration/com/thoughtworks/go/server/service/UserServiceIntegrationTest.java [hunk 3]
```diff
@@ -609,6 +611,73 @@ public void shouldDisableAdminSelectionWhenUserIsMemberOfAdminRole() throws Exce
...
+        userDao.saveOrUpdate(new User("admin-user"));
+        userDao.saveOrUpdate(new User("user-1"));
+        userService.enable(Arrays.asList("user-1"), new HttpLocalizedOperationResult());
+
+        HttpLocalizedOperationResult result = new HttpLocalizedOperationResult();
+        userService.update(new User("user-1"), TriState.FALSE, TriState.UNSET, null, null, result);
```
Comment (zabil) [inline server/test/integration/com/thoughtworks/go/server/service/UserServiceIntegrationTest.java]: Ok, I am not user we should pass the user object over here.
We are are re-assigning this parameter in the service. Why can't we just take in the user name.
We don't reload the user name in the service.

File: server/src/com/thoughtworks/go/server/service/UserService.java [hunk 4]
```diff
@@ -131,6 +120,40 @@ private boolean userNameListContainsAdmin(List<String> enabledUserNames) {
...
+    public void update(User user, TriState enabled, TriState emailMe, String email, String checkinAliases, LocalizedOperationResult result) {
+        if (enabled.isTrue()) {
+            enable(Arrays.asList(user.getName()), result);
+        }
+
+        if (enabled.isFalse()) {
+            disable(Arrays.asList(user.getName()), result);
+        }
+
+        user = findUserByName(user.getName());
```
Comment (zabil) [inline server/src/com/thoughtworks/go/server/service/UserService.java]: Prefer us not doing this as the controller has loaded the user.
Or change the method to take in the username and then do the load.

Comment (ketan) [inline server/src/com/thoughtworks/go/server/service/UserService.java]: We're having reload the user because the enable/disable happens in a transaction of its own, the other updates happen in a different transaction.

I'm happy to hear how to fix this properly though.

File: server/test/integration/com/thoughtworks/go/server/service/UserServiceIntegrationTest.java (hunk 3 above)
Comment (ketan) [inline server/test/integration/com/thoughtworks/go/server/service/UserServiceIntegrationTest.java]: Can I change the implementation to this -

```java
    public User update(User user, TriState enabled, TriState emailMe, String email, String checkinAliases, LocalizedOperationResult result) {
        if (enabled.isTrue()) {
            user.enable();
        }

        if (enabled.isFalse()) {
            user.disable();
        }

        if (email != null) {
            user.setEmail(email);
        }

        if (checkinAliases != null) {
            user.setMatcher(checkinAliases);
        }

        if (emailMe.isTrue()) {
            user.setEmailMe(true);
        }

        if (emailMe.isFalse()) {
            user.setEmailMe(false);
        }

        try {
            saveOrUpdate(user);
        } catch (ValidationException e) {
            result.badRequest(LocalizedMessage.string("USER_FIELD_VALIDATIONS_FAILED", e.getMessage()));
        }

        return user;
    }
```

Notice that it updates and mutates the same object and attempts to to save it in a single transaction.

This prevents a double load from the controller as well as from inside the service. The service does exactly what it advertises "update", no more.

Comment (zabil) [inline server/test/integration/com/thoughtworks/go/server/service/UserServiceIntegrationTest.java]: Yes this makes sense.
I'd also make the user param final because this way of saving the user sadly unconventional because of our jruby way of doing things .

File: server/test/integration/com/thoughtworks/go/server/service/UserServiceIntegrationTest.java [hunk 5]
```diff
@@ -609,6 +611,73 @@ public void shouldDisableAdminSelectionWhenUserIsMemberOfAdminRole() throws Exce
...
+        userService.update(new User("user-1"), TriState.UNSET, TriState.UNSET, "foo@example.com", null, result);
+        assertThat(result.isSuccessful(), is(true));
+
+        assertThat(userService.findUserByName("user-1").getEmail(), is("foo@example.com"));
+
+        result = new HttpLocalizedOperationResult();
+        userService.update(new User("user-1"), TriState.TRUE, TriState.UNSET, "", null, result);
+        assertThat(userService.findUserByName("user-1").getEmail(), is(""));
+        assertThat(result.isSuccessful(), is(true));
+    }
```
Comment (zabil) [inline server/test/integration/com/thoughtworks/go/server/service/UserServiceIntegrationTest.java]: The new line police is upset.

**Comment by ketan** [comment]
@zabil - addressed all your feedback. Ready to merge. Also rebased the PR for backup api on top of this.

File: server/test/integration/com/thoughtworks/go/server/service/UserServiceIntegrationTest.java [hunk 6]
```diff
@@ -609,6 +611,73 @@ public void shouldDisableAdminSelectionWhenUserIsMemberOfAdminRole() throws Exce
...
+
+    @Test
+    public void updateShouldUpdateEmailMeState() throws Exception {
+        userDao.saveOrUpdate(new User("user-1"));
+
+        User user = userService.findUserByName("user-1");
+        user.setEmailMe(false);
+        userDao.saveOrUpdate(user);
+
+        userService.enable(Arrays.asList("user-1"), new HttpLocalizedOperationResult());
```
Comment (zabil) [inline server/test/integration/com/thoughtworks/go/server/service/UserServiceIntegrationTest.java]: Should we enable the user for an update?

I think there are too many asserts in the test cases can't we split it up like this?

```
    @Test
    public void shouldUpdateEmailMeStateToTrue() throws Exception {
        User user = new User("user-1");
        user.setEmailMe(false);
        addUser(user);

        HttpLocalizedOperationResult result = new HttpLocalizedOperationResult();
        userService.update(user, TriState.UNSET, TriState.TRUE, null, null, result);

        assertThat(result.isSuccessful(), is(true));
        assertThat(userService.findUserByName("user-1").isEmailMe(), is(true));

    }
    @Test
    public void shouldUpdateEmailMeStateToFalse() throws Exception {
        User user = new User("user-1");
        user.setEmailMe(false);
        addUser(user);

        HttpLocalizedOperationResult result = new HttpLocalizedOperationResult();
        userService.update(user, TriState.UNSET, TriState.FALSE, null, null, result);

        assertThat(result.isSuccessful(), is(true));
        assertThat(userService.findUserByName("user-1").isEmailMe(), is(false));
    }
```

Also use the ```addUser``` method

Comment (ketan) [inline server/test/integration/com/thoughtworks/go/server/service/UserServiceIntegrationTest.java]: Technically it's a tristate, we'd need 3 different tests for each of the tristates, plus 2 additional tests for EACH for email and checkinAliases (with null values and without null) I feel these tests are being too verbose, and perhaps don't as much value as I'd thought.

WDYT?

Comment (zabil) [inline server/test/integration/com/thoughtworks/go/server/service/UserServiceIntegrationTest.java]: I know,
 But I found it tough to read one method doing a lot of operations.

 I feel it's cleaner to have one method call and one assertion for the
result of that method call. So that I can easily map what the assertion is
testing.

File: server/test/integration/com/thoughtworks/go/server/service/UserServiceIntegrationTest.java [hunk 7]
```diff
@@ -341,7 +337,7 @@ public void shouldEnableUsers() throws Exception {
         HttpLocalizedOperationResult result = new HttpLocalizedOperationResult();
```
Comment (zabil) [inline server/test/integration/com/thoughtworks/go/server/service/UserServiceIntegrationTest.java]: Noticed that the 
```
userService.enable(Arrays.asList("user_one"), new HttpLocalizedOperationResult());
```

On this method takes a new instance of ```new HttpLocalizedOperationResult()``` it should be taking in ```result``` instead.

Comment (ketan) [inline server/test/integration/com/thoughtworks/go/server/service/UserServiceIntegrationTest.java]: That is not my change :)

Comment (zabil) [inline server/test/integration/com/thoughtworks/go/server/service/UserServiceIntegrationTest.java]: True,
but a broken window ;-)

Comment (ketan) [inline server/test/integration/com/thoughtworks/go/server/service/UserServiceIntegrationTest.java]: Can we call this done now?

**Comment by zabil** [comment]
Verified on experimental release15.2.0-2219 (dae53017b0071763841d23c48a0c1d5e29dc661e)
